design/disaster-recovery-snapshots.md

```diff
@@ -0,0 +1,128 @@
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/kotsadm: "true"
+
+- Application restore will be created with `labelSelector` set as in the example below:
+
+        apiVersion: velero.io/v1
+        kind: Restore
+        spec:
+          labelSelector:
+            matchLabels:
+              kots.io/app-slug: <slug>
+
+- Restore can be performed using Velero CLI:
+
+        velero restore create --from-backup <backup name> -l kots.io/backup=velero
+
+        velero restore create --from-backup <backup name> -l kots.io/app-slug=<slug>
+
+- `kots install --from-backup` command will be implemented as a wrapper for the above commands
+    - `kots restore` is a potential alternative, but the UI-based approached discussed next will also need a `kots install` command.
+
+
+### Potential blockers
+
+When adding new labels, the next deployment can potentially fail with the `field is immutable` errors.
+A workaround is to delete the asset that causes the error prior to deploying the new version.
+
+### UI-based restores
+
+In order to allow a UI-based disaster recovery flow, Admin Console needs to be installed first.
+When restoring the Admin Console resources, `kotsadm` pod will be restarted leaving UI in a potentially bad state.
+For example, once data is restored, the logged in user's session will be invalidated.
+Also receiving restore status will be impossible if the API is not available or session is invalid.
+
+
+In order to avoid the above problems, the following can be implemented:
+
+
+1. Only Postgres data, configmaps, and secrets will be backed up (i.e. pods will not have the `kots.io/backup=velero` label).
+1. There will be a sidecar pod created (`kotsadm-backup`) that will have the `kots.io/backup=velero` label
+    - `kotsadm-backup` will run pre-snapshot hooks to backup Postgres data.
+    - `kotsadm-backup` will run post-restore hook to restore Postgres data
+    - `kotsadm-backup` will be deleted once backup or restore has completed.
+1. `kots install` command will allow installing Admin Console only.
+
+
+Because Admin Console pods will not be restored, we need to ensure that the version installed is compatible with the the data stored in the snapshot.
+To accomplish this, the following will be implemented:
+
+
+1. All Velero backup objects will have the `kots.io/version` label.
+1. During the restore, this label will be used to filter out backups that have non-matching versions.
+1. During the restore, users will be informed that a different version of Admin Console needed if they want to restore a different, incompatible backup.
+
+
+### kURL registry
+
+Airgapped embedded instances will also have images stored in local kURL registry.
+These images can be backed up and restored using the `kotsadm-backup` sidecar pod.
+
+## Alternatives Considered
+
+Storing Admin Console in its own separate snapshot has been considered.
+
+## Security Considerations
+
+None
```

Comment:
Some questions:

I'd like some detail on how we expect snapshots to be configured. Should they be configured at the instance level or at the app level? I believe today it's a big half-and-half. If someone triggers a backup of one app, should all apps be backed up? If so, treating this as an instance-level snapshot may make more sense. Just like most backup systems, you back up the whole thing, even if you only want to restore one thing later. 

In the future, for a really large enterprise user with 10,000 applications, that may be a big cumbersome. But for all practical purposes today, it would be fine.

But being able to trigger an app-level snapshot is nice as it's common to do right before making a major change.

So, if we perpetuate a prior design where triggering an application-specific snapshot also snapshots the admin console, but none of the other apps, what would we do if someone restores that admin console snapshot? Would we only restore the one app saved with it? Or would we dig for a prior snapshot that restores the rest of the apps? Even if that all works, it would be very hard for end-users to understand. It may also be dangerous in that Admin Console will show a record of a certain version running which may be different than was restored. Even if we do the magic adjustment to point to the right place, it might be confusing to the user why their app lost a version. For disaster recovery, that may be better than nothing. And it depends how old the backup us.

Maybe we shouldn't let people disaster recovery restore from a partial snapshot. Treat that like the intent it had, which was an app-level rollback mechanism. Maybe even call them different things. Two different features, app snapshots and cluster backups (although that name is misleading...).

Or, just have every snapshot trigger cluster-wide snapshot.

For registry backup, should we support restoring to a different place? One of the use cases people want is to use backup/restore not just to recover to a new cluster, but to recover to a different *type* of cluster. Maybe start with a single-node kURL cluster, then restore to a *real* cluster. That could be a multi-node HA cluster, but it may be EKS with the registry as ECR.

Does this help us at all towards being able to move apps from one multi-app cluster to another? e.g. After restore, trigger an internal delete of the records for the apps not copied over?